### PR TITLE
Added support for most RSA compatible ciphers

### PR DESCRIPTION
**Pull request by alexmgr**
- Refactored TLSSecurityParameters to handle most ciphers
- Refactored TLSSessionCtx to use the new security parameters
- Added TLSCompressionParameters to deal with compression
- Wrote unit tests for above classes
- Wrote a pycrypto compatible implementation of a Null cipher
and Null hash
- Wrote a zlib compatible implementation of Null compression
- Removed cipher_factory() and describe_cipher()
- Removed PKCS7 automatic encoding. Will move this into the
crypto helper classes in another commit. Will allow side channel
testing

I refactored this layer quite heavily to allow easy support for all cipher and compression methods. I also tried to group all necessary parameters inside the TLSSessionCtx as you started.
The plan is never to use individual parameters, but just pass handles to TLSSessionCtx to functions, which can then do encryption/compression/... transparently.

I don't think this conflicts too much with the tlsfinished branch, and should be able to integrate fairly nicely pending a bit of refactoring of the function definitions.

### CODE REVIEW

File: src/scapy/layers/ssl_tls_crypto.py [hunk 1]
```diff
@@ -36,21 +35,10 @@ def prf(master_secret, label, data):
 def x509_extract_pubkey_from_der(der_certificate):
     # Extract subjectPublicKeyInfo field from X.509 certificate (see RFC3280)
     cert = DerSequence()
-    cert.decode(der_certificate)                
-    
+    cert.decode(der_certificate)
     tbsCertificate = DerSequence()
-    tbsCertificate.decode(cert[0])       # first DER SEQUENCE
```
Comment (tintinweb) [inline src/scapy/layers/ssl_tls_crypto.py]: reverts 048c353eb23cbf875327c3da973a340cba3c5362
relying on tbscertificate[6] is error prone; the fix from 048c353eb23cbf875327c3da973a340cba3c5362 does a quick search for the OID 'rsaEncryption'

File: src/scapy/layers/ssl_tls_crypto.py [hunk 2]
```diff
@@ -465,116 +415,213 @@ def hmac(self, key, msg):
...
-    pre_master_secret = "\03\01aaaaaaaaaaaaaaaaaaaaaabbbbbbbbbbbbbbbbbbbbbbbb"
-    client_random = 'a'*32
-    server_random = 'z'*32
+class UnsupportedCipherError(Exception):
+    pass
+ 
+class TLSSecurityParameters(object):
+    
+    crypto_params = {
+                    0x0000: {"name":"NULL_WITH_NULL_NULL", "export":False, "key_exchange":{"type":RSA, "name":"RSA"}, "cipher":{"type":NullCipher, "name":"Null", "key_len":0, "mode":None, "mode_name":""}, "hash":{"type":NullHash, "name":"Null"}},
```
Comment (tintinweb) [inline src/scapy/layers/ssl_tls_crypto.py]: we might want to reuse the TLS_CIPHER_SUITES enum from ssl_tls.py

File: src/scapy/layers/ssl_tls_crypto.py [hunk 3]
```diff
@@ -465,116 +415,213 @@ def hmac(self, key, msg):
...
+        return data
+
+    @staticmethod
+    def decompress(data):
+        return data
+
+class TLSCompressionParameters(object):
     
+    comp_params = {
+                  0x00: {"name":"Null", "type":NullCompression},
```
Comment (tintinweb) [inline src/scapy/layers/ssl_tls_crypto.py]: we might want to reuse the TLS_COMPRESSION_METHODS enum from ssl_tls.py here

File: src/scapy/layers/ssl_tls_crypto.py (hunk 1 above)
Comment (alexmgr) [inline src/scapy/layers/ssl_tls_crypto.py]: Good catch. I didn't intend to remove that.
I'll add that back in.

File: src/scapy/layers/ssl_tls_crypto.py (hunk 2 above)
Comment (alexmgr) [inline src/scapy/layers/ssl_tls_crypto.py]: Agreed

File: src/scapy/layers/ssl_tls_crypto.py (hunk 3 above)
Comment (alexmgr) [inline src/scapy/layers/ssl_tls_crypto.py]: Agreed

**Comment by tintinweb** [comment]
excellent, thanks! :+1:

**Pull request merged**